File: array/182.py
from typing import List
import heapq
import math


def pickGifts(gifts: List[int], k: int) -> int:
    l1 = [-num for num in gifts]
    heapq.heapify(l1)

    while k > 0:
        elem = -heapq.heappop(l1)
        heapq.heappush(l1, -math.floor(math.sqrt(elem)))
        k -= 1

    return abs(sum(l1))


print(
    pickGifts([25, 64, 9, 4, 1000000000, 1000000000, 999999999, 12232323, 3434344], 4)
)

File: array/182.cpp
#include <bits/stdc++.h>

using namespace std;

long long pickGifts(vector<int>& gifts, int k) {
    priority_queue<int>pq;
    vector<int> ans;

    for(const auto& value: gifts)
        pq.push(value);

    while(k-- > 0) {
        int elem = pq.top();
        pq.pop();
        int squareRoot = static_cast<int>(sqrt(elem));
        pq.push(squareRoot);
    }

    long long int sum = 0;

    while(!pq.empty()) {
        sum += pq.top();
        pq.pop();
    }

    return sum;
}

int main() {
    vector<int> nums = {1,1,1,1};
    cout<<pickGifts(nums, 4);

    return 0;
}